filebeat/prospector/log/harvester.go

```diff
@@ -111,3 +125,381 @@ func (h *Harvester) open() error {
+func (h *Harvester) Stop() {
+	h.stop()
+	h.stopWg.Wait()
+}
+
+// sendEvent sends event to the spooler channel
+// Return false if event was not sent
+func (h *Harvester) sendEvent(data *util.Data) bool {
+	if h.source.HasState() {
+		h.states.Update(data.GetState())
+	}
+
+	err := h.forwarder.Send(data)
+	return err == nil
+}
+
+// sendStateUpdate send an empty event with the current state to update the registry
+// close_timeout does not apply here to make sure a harvester is closed properly. In
+// case the output is blocked the harvester will stay open to make sure no new harvester
+// is started. As soon as the output becomes available again, the finished state is written
+// and processing can continue.
+func (h *Harvester) SendStateUpdate() {
+
+	if !h.source.HasState() {
+		return
+	}
+
+	logp.Debug("harvester", "Update state: %s, offset: %v", h.state.Source, h.state.Offset)
+	h.states.Update(h.state)
+
+	d := util.NewData()
+	d.SetState(h.state)
+	h.forwarder.Outlet.OnEvent(d)
+}
+
+// shouldExportLine decides if the line is exported or not based on
+// the include_lines and exclude_lines options.
+func (h *Harvester) shouldExportLine(line string) bool {
+	if len(h.config.IncludeLines) > 0 {
+		if !harvester.MatchAny(h.config.IncludeLines, line) {
+			// drop line
+			logp.Debug("harvester", "Drop line as it does not match any of the include patterns %s", line)
+			return false
+		}
+	}
+	if len(h.config.ExcludeLines) > 0 {
+		if harvester.MatchAny(h.config.ExcludeLines, line) {
+			// drop line
+			logp.Debug("harvester", "Drop line as it does match one of the exclude patterns%s", line)
+			return false
+		}
+	}
+
+	return true
+
+}
+
+// openFile opens a file and checks for the encoding. In case the encoding cannot be detected
+// or the file cannot be opened because for example of failing read permissions, an error
+// is returned and the harvester is closed. The file will be picked up again the next time
+// the file system is scanned
+func (h *Harvester) openFile() error {
+
+	f, err := file.ReadOpen(h.state.Source)
+	if err != nil {
+		return fmt.Errorf("Failed opening %s: %s", h.state.Source, err)
+	}
+
+	harvesterOpenFiles.Add(1)
+
+	// Makes sure file handler is also closed on errors
+	err = h.validateFile(f)
+	if err != nil {
+		f.Close()
+		harvesterOpenFiles.Add(-1)
+		return err
+	}
+
+	h.source = source.File{File: f}
+	return nil
+}
+
+func (h *Harvester) validateFile(f *os.File) error {
+
+	info, err := f.Stat()
+	if err != nil {
+		return fmt.Errorf("Failed getting stats for file %s: %s", h.state.Source, err)
+	}
+
+	if !info.Mode().IsRegular() {
+		return fmt.Errorf("Tried to open non regular file: %q %s", info.Mode(), info.Name())
+	}
+
+	// Compares the stat of the opened file to the state given by the prospector. Abort if not match.
+	if !os.SameFile(h.state.Fileinfo, info) {
+		return errors.New("File info is not identical with opened file. Aborting harvesting and retrying file later again.")
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
error strings should not be capitalized or end with punctuation or a newline